### PR TITLE
NodeGraph exposure with test cases

### PR DESCRIPTION
**Pull request by fujitatomoya**
address https://github.com/ros2/rclcpp/issues/1483

### CODE REVIEW

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 1]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
     const std::string & node_name,
     const std::string & namespace_) const = 0;
 
+  /// Return a map of existing service client names and types with a specific node.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:186]:
```suggestion
  /// Return a map of existing service names and types with a specific node.
```

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 2]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+   * \param[in] namespace_ namespace of the node
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::map<std::string, std::vector<std::string>>
+  get_client_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_) const = 0;
+
+  /// Return a map of existing publisher topic names to list of topic types for a specific node.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:201]:
```suggestion
  /// Return a map of existing topic names to list of topic types for a specific node.
```

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 3]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+  /**
+   * This function only considers publications - not subscriptions.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:203]:
```suggestion
   * This function only considers publishers - not subscribers.
```

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 4]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::map<std::string, std::vector<std::string>>
+  get_publisher_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_,
+    bool no_demangle = false) const = 0;
+
+  /// Return a map of existing subscriber topic names to list of topic types for a specific node.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:218]:
```suggestion
  /// Return a map of existing topic names to list of topic types for a specific node.
```

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 5]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+  /**
+   * This function only considers subscriptions - not publications.
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:220]:
```suggestion
   * This function only considers subscribers - not publishers.
```

**Comment by fujitatomoya** [comment]
friendly ping @ivanpauno @clalancette

**Comment by fujitatomoya** [comment]
@mabelzhang friendly ping

**Comment by fujitatomoya** [comment]
@mabelzhang friendly ping for another review.

**Comment by fujitatomoya** [comment]
@ivanpauno @clalancette friendly ping.

**Comment by fujitatomoya** [comment]
CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14072/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8891/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11764/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14164/)

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 6]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+   * \param[in] namespace_ namespace of the node
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::map<std::string, std::vector<std::string>>
+  get_client_names_and_types_by_node(
+    const std::string & node_name,
+    const std::string & namespace_) const = 0;
+
+  /// Return a map of existing topic names to list of topic types for a specific node.
```
Comment (iuhilnehc-ynos) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:201]: I don't know if one topic name can support multiple types in a node. (Actually, I am afraid not.)
https://github.com/ros2/rmw/blob/ace271650ad39a656d81239050c9639c4190f077/rmw/include/rmw/names_and_types.h#L37

Comment (fujitatomoya) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:201]: thanks, related to https://github.com/ros2/ros2/issues/1095.

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp [hunk 7]
```diff
@@ -178,6 +179,136 @@ NodeGraph::get_service_names_and_types_by_node(
   return services_and_types;
 }
 
+std::map<std::string, std::vector<std::string>>
+NodeGraph::get_client_names_and_types_by_node(
+  const std::string & node_name,
+  const std::string & namespace_) const
+{
+  rcl_names_and_types_t service_names_and_types = rcl_get_zero_initialized_names_and_types();
+  rcl_allocator_t allocator = rcl_get_default_allocator();
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:188]: It would be good if an allocator could be provided somehow.
Though it seems a bit unrelated to the PR.

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp [hunk 8]
```diff
@@ -178,6 +179,136 @@ NodeGraph::get_service_names_and_types_by_node(
...
+      rcl_get_error_string().str;
+    rcl_reset_error();
+    if (rcl_names_and_types_fini(&service_names_and_types) != RCL_RET_OK) {
+      error_msg +=
+        std::string(", failed also to cleanup service names and types, leaking memory: ") +
+        rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    throw std::runtime_error(error_msg);
+  }
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:206]: Maybe using [throw_from_rcl_error](https://github.com/ros2/rclcpp/blob/63e79223f98867ba00c059eb612438ca9461a2f7/rclcpp/include/rclcpp/exceptions/exceptions.hpp#L128) and [rcpputils::make_scope_exit](https://github.com/ros2/rcpputils/blob/bc300e7667603a4b8a6cc2b9d7b13f5b64e79265/include/rcpputils/scope_exit.hpp#L58) (?)

something like:

```cpp
  rcl_names_and_types_t service_names_and_types = rcl_get_zero_initialized_names_and_types();
  auto service_names_and_types_finalizer = rcpputils::make_scope_exit([&service_names_and_types] () {
    if (rcl_names_and_types_fini(&service_names_and_types) != RCL_RET_OK) {
      RCLCPP_ERROR(....);
      rcl_reset_error();
    }
  });
  rcl_ret_t ret = rcl_get_client_names_and_types_by_node(...);
  if (ret != RCL_RET_OK) {
    rclcpp::exception::throw_from_rcl_error(...);
  }
  ...
  service_names_and_types_finalizer.cancel();
```

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp [hunk 9]
```diff
@@ -178,6 +179,136 @@ NodeGraph::get_service_names_and_types_by_node(
...
+    std::string service_name = service_names_and_types.names.data[i];
+    for (size_t j = 0; j < service_names_and_types.types[i].size; ++j) {
+      services_and_types[service_name].emplace_back(service_names_and_types.types[i].data[j]);
+    }
+  }
+
+  ret = rcl_names_and_types_fini(&service_names_and_types);
+  if (ret != RCL_RET_OK) {
+    throw_from_rcl_error(ret, "could not destroy service names and types");
+  }
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:219]: following the comment above, this won't be needed

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp (hunk 8 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:206]: similar comment for the other implementations

**Comment by fujitatomoya** [comment]
@ivanpauno thanks for the comments, i will check and address them.

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 10]
```diff
@@ -182,6 +183,55 @@ class NodeGraphInterface
...
+  /**
+   * This function only considers publishers - not subscribers.
+   * The returned names are the actual names used and do not have remap rules applied.
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:204]: right, that is confusing for me too.

probably `rcl/graph` and `rcl_action/graph` comments are also confusing?

https://github.com/ros2/rcl/blob/b7784eb5d2fbff33053b69f6b9f9d3b703f9f2a7/rcl/include/rcl/graph.h#L76-L78

https://github.com/ros2/rcl/blob/b7784eb5d2fbff33053b69f6b9f9d3b703f9f2a7/rcl_action/include/rcl_action/graph.h#L38-L40

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp (hunk 7 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:188]: probably, we can have `allocator` in `NodeGraph` class. let's skip this one since it is not really related to the fix.

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp (hunk 6 above)
Comment (iuhilnehc-ynos) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:201]: @ivanpauno 

> but you cannot control what other nodes in the graph do.

Sorry, I don't catch it. Could you give me more information about what other nodes could do in the graph?

all DDS vendor will use  to get names and types from
https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/src/graph_cache.cpp#L652

no idea why using the [std::set < std::string >](https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/src/graph_cache.cpp#L637) here, we can see the [demangle_type(item.second.topic_type)](https://github.com/ros2/rmw_dds_common/blob/9124fc6b915064e171ba24c722be7c9d7fefcecb/rmw_dds_common/include/rmw_dds_common/graph_cache.hpp#L345) can only return `std::string`.

It seems we expect one topic name can only have one relative topic type, if so, why do we need to use `vector` there?

Comment (iuhilnehc-ynos) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:201]: Sorry for the noise.

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp (hunk 8 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:206]: i think that would be nicer. borrowing your pseudo code.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14132/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8937/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11819/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14230/)

**Comment by fujitatomoya** [comment]
https://ci.ros2.org/job/ci_windows/14230/cmake/new/ is unrelated.

**Comment by fujitatomoya** [comment]
i did some mis-operation to overwrite the fix... 😢 i will re-check again.

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:204]: > probably rcl/graph and rcl_action/graph comments are also confusing?

What that means is that you're getting that already remapped name, but if you pass that to `create_publisher()` the result might be wrong because of remapping rules being applied again.

The ` The returned names are not automatically remapped by this function.` makes the comment a bit confusing, so if you want to rephrase that that would be ok.
But the second one is the really important one.

P.S.: We should actually have a way to create a topic with **exactly** a name, which isn't currently possible.

**Review by ivanpauno — CHANGES_REQUESTED** [review]
Sorry, but we're past the API/feature freeze so I will hold this PR until we create a `galactic` branch.

**Comment by fujitatomoya** [comment]
no problem, i will keep this to myself to next release then.

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:204]: > But the second one is the really important one.

okay, i will add comments about this `do not have remap rules applied` part.

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14282/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9071/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11966/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14382/)

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14646/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9399/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12310/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14790/)

**Comment by ivanpauno** [comment]
@fujitatomoya friendly ping.
Is this ready for review again?

**Comment by fujitatomoya** [comment]
@ivanpauno i will rebase and check, and then get back to you.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15018/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9757/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12707/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15180/)

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

**Comment by fujitatomoya** [comment]
@ivanpauno could you take a look when you have time? thanks for ping!

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 11]
```diff
@@ -146,7 +147,9 @@ class NodeGraphInterface
   /**
    * A topic is considered to exist when at least one publisher or subscriber
    * exists for it, whether they be local or remote to this process.
-   * The returned names are the actual names used and do not have remap rules applied.
+   * The returned names are the actual names after remap rules applied.
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:150]: I think I suggested this change because I got confused with the original wording, e.g.:
```
node->create_publisher("my_topic", ...)
auto result = another_node->get_topic_names_and_types();
```
The result will have an item `/first/node/namespace/my_topic`, and not simply `my_topic`.
But now I see that the original comment was about `another_node` not transforming the discovered topic name at all, which is also true (remapping rules are applied by `node`, not by `another_node`).

See also https://github.com/ros2/rclcpp/pull/1225, https://github.com/ros2/rclcpp/issues/1174.

Maybe we could reword this as:

```
The returned names are the actual names of the topics, either announced by another nodes or by this one.
No transformation is applied to the announced names by this node.
```

@sloretz what do you think? (I think you introduced https://github.com/ros2/rclcpp/pull/1225).

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp [hunk 12]
```diff
@@ -206,6 +317,85 @@ NodeGraph::get_node_names() const
...
+      error_msg += std::string(", failed also to cleanup node namespaces, leaking memory: ") +
+        rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    if (rcutils_string_array_fini(&node_enclaves_c) != RCUTILS_RET_OK) {
+      error_msg += std::string(", failed also to cleanup node enclaves, leaking memory: ") +
+        rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    // TODO(karsten1987): Append rcutils_error_message once it's in master
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:356]: does this comment still apply?

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp [hunk 13]
```diff
@@ -206,6 +317,85 @@ NodeGraph::get_node_names() const
...
+    if (node_names_c.data[i] && node_namespaces_c.data[i] && node_enclaves_c.data[i]) {
+      node_tuples.emplace_back(
+        std::make_tuple(node_names_c.data[i], node_namespaces_c.data[i], node_enclaves_c.data[i]));
+    }
+  }
+
+  std::string error;
+  rcl_ret_t ret_names = rcutils_string_array_fini(&node_names_c);
+  if (ret_names != RCUTILS_RET_OK) {
+    // *INDENT-OFF*
```
Comment (ivanpauno) [inline rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:371]: are the `*INDENT-OFF*` lines needed?
I think those were copy-pasted, but I'm not sure if they still apply.

File: rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp [hunk 14]
```diff
@@ -259,12 +312,163 @@ TEST_F(TestNodeGraph, get_service_names_and_types_by_node_names_and_types_fini_e
...
+
+  // Check that node1_topic exists for node1 but not node2. This shouldn't exercise graph
+  // discovery as node_graph belongs to node1 anyway. This is just to test the API itself.
+  auto topics_of_node1 =
+    node_graph()->get_publisher_names_and_types_by_node(node_name, absolute_namespace);
+  auto topics_of_node2 =
+    node_graph()->get_publisher_names_and_types_by_node(node2_name, absolute_namespace);
+
+  auto start = std::chrono::steady_clock::now();
+  while (std::chrono::steady_clock::now() - start < std::chrono::seconds(1)) {
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:346]: can we give this a timeout larger than 1 second?
I can already imagine tests failing with rmw_connextdds

Comment (ivanpauno) [reply rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:346]: same in other places

**Review by ivanpauno — APPROVED** [review]

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp (hunk 11 above)
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp:150]: @ivanpauno @sloretz thanks for the consideration, sounds good to me. i will update it accordingly.

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp (hunk 12 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:356]: I think it is already appended.

File: rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp (hunk 14 above)
Comment (fujitatomoya) [reply rclcpp/test/rclcpp/node_interfaces/test_node_graph.cpp:346]: change this into 3 seconds.

File: rclcpp/src/rclcpp/node_interfaces/node_graph.cpp (hunk 13 above)
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/node_interfaces/node_graph.cpp:371]: nice catch, actually we do not need this.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15104/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9836/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12791/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15272/)

File: rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp [hunk 15]
```diff
@@ -182,18 +189,82 @@ class NodeGraphInterface
...
   get_node_names() const = 0;
 
+  /// Return a vector of existing node names, namespaces and enclaves (tuple of string).
+  /*
+   * The returned names are the actual names after remap rules applied.
+   */
+  RCLCPP_PUBLIC
+  virtual
+  std::vector<std::tuple<std::string, std::string, std::string>>
+  get_node_names_with_enclaves() const = 0;
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp]: maybe, we can add a link to https://design.ros2.org/articles/ros2_security_enclaves.html

[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/node_interfaces/node_graph_interface.hpp]: that is good idea, i will change it accordingly.

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

**Comment by fujitatomoya** [comment]
@ros-pull-request-builder retest this please

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15245/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9964/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12926/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15431/)

**Comment by fujitatomoya** [comment]
CI unstable, retry:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15263/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9979/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12944/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15456/)

**Comment by fujitatomoya** [comment]
I am not sure why this is failing, retry.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15271/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9987/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12953/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15464/)

**Comment by clalancette** [comment]
> I am not sure why this is failing, retry.

Sorry, we had a bug in the CI scripts that was fixed in https://github.com/ros2/ci/pull/592 .  You probably got unlucky and started your jobs between when we merged the bug and merged the fix.  Hopefully your latest try will work :crossed_fingers:

**Comment by fujitatomoya** [comment]
CI still fails, according to https://ci.ros2.org/job/ci_linux/15279/console, i guess that i need to rebase on latest [ros2.repo](https://raw.githubusercontent.com/ros2/ros2/master/ros2.repos) to change the branch for cyclonedds.

**Comment by fujitatomoya** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15281/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9995/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12962/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15473/)
